framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1359,13 +1359,25 @@ public Void visitUnary(UnaryTree node, Void p) {
                 || (node.getKind() == Tree.Kind.POSTFIX_DECREMENT)
                 || (node.getKind() == Tree.Kind.POSTFIX_INCREMENT)) {
             AnnotatedTypeMirror varType = atypeFactory.getAnnotatedTypeLhs(node.getExpression());
-            // For postfix increments/decrements, the value type is incorrect due to the workaround
-            // in GenericAnnotatedTypeFactory.addComputedTypeAnnotations(Tree, AnnotatedTypeMirror, boolean)
-            // for the following bug:
-            // See Issue 867: https://github.com/typetools/checker-framework/issues/867
-            // This means could result in a false warning (false positive) in some cases and a lack
-            // of a warning in other cases (false negative).
-            AnnotatedTypeMirror valueType = atypeFactory.getAnnotatedType(node);
+            Tree newValue = node;
+            AnnotatedTypeMirror valueType = atypeFactory.getAnnotatedType(newValue);
+
+            // When the operation is postfix, valueType is a type of a value before the
+            // assignment. We override valueType with a type from a store, if it's provided.
+            if (node.getKind() == Tree.Kind.POSTFIX_DECREMENT
+                    || node.getKind() == Tree.Kind.POSTFIX_INCREMENT) {
+                CFAbstractStore<?, ?> store = atypeFactory.getStoreAfter(newValue);
+                if (store != null) {
+                    CFAbstractValue<?> value =
```

Comment:
I'm wondering whether we should do something similar to boxing/unboxing of primitives.
If we do the special treatment in the Visitor, everybody who uses the CFG needs to be aware of the distinction and implement it.
For boxing/unboxing we keep a reference to the Tree before conversion. See Section 3.3.2 in the dataflow manual and https://github.com/typetools/checker-framework/blob/master/dataflow/src/org/checkerframework/dataflow/cfg/ControlFlowGraph.java#L53

Can you argue why unary increment/decrement shouldn't be handled similarly?
This reminds me of another thing: this PR should also update the dataflow manual and document what tree transformation we're performing.